dsls/sql/src/main/java/org/apache/beam/dsls/sql/BeamSql.java

```diff
@@ -93,42 +98,62 @@
    *
    * <p>Make sure to query it from a static table name <em>PCOLLECTION</em>.
    */
-  public static PTransform<PCollection<BeamSqlRow>, PCollection<BeamSqlRow>>
-  simpleQuery(String sqlQuery) throws Exception {
-    return new SimpleQueryTransform(sqlQuery);
+  public static SimpleQueryTransform simpleQuery(String sqlQuery) throws Exception {
+    return SimpleQueryTransform.builder()
+        .setSqlEnv(new BeamSqlEnv())
+        .setSqlQuery(sqlQuery)
+        .build();
   }
 
   /**
    * A {@link PTransform} representing an execution plan for a SQL query.
    */
-  private static class QueryTransform extends
+  @AutoValue
+  public abstract static class QueryTransform extends
       PTransform<PCollectionTuple, PCollection<BeamSqlRow>> {
-    private transient BeamSqlEnv sqlEnv;
-    private String sqlQuery;
+    abstract BeamSqlEnv getSqlEnv();
+    abstract String getSqlQuery();
 
-    public QueryTransform(String sqlQuery) {
-      this.sqlQuery = sqlQuery;
-      sqlEnv = new BeamSqlEnv();
+    static Builder builder() {
+      return new AutoValue_BeamSql_QueryTransform.Builder();
     }
 
-    public QueryTransform(String sqlQuery, BeamSqlEnv sqlEnv) {
-      this.sqlQuery = sqlQuery;
-      this.sqlEnv = sqlEnv;
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setSqlQuery(String sqlQuery);
+      abstract Builder setSqlEnv(BeamSqlEnv sqlEnv);
+      abstract QueryTransform build();
     }
 
+    /**
+     * register a UDF function used in this query.
+     */
+     public QueryTransform withUdf(String functionName, Class<?> clazz, String methodName){
```

Comment:
1. Is it necessary to support having arbitrarily named (and potentially multiple) UDF methods on a single class? 
[Mingmin]:  I would like to keep the capacity to put similar or related methods together.

2. Why not just have a simply BeamSqlUdf interface you must implement? 
[Mingmin]: That's how we do with UDAF, UDF is a little different as it may have variable, even optional parameters. Annotation (like `DoFn`) sounds a solution but the existing definition is more simple IMO.

3. And why not accept a lambda?
[Mingmin]: I remember Beam is still on JDK 7, --sorry that I may miss something.